### ISSUE TITLE
Pause microphone while audio is playing back? (Javascript)

### ISSUE THREAD

**Issue by Marcophono2** [Azure-Samples/cognitive-services-speech-sdk#171]
Hello!
I am using the demo code from the Javascript version for translation of microphone input into another language. I coded a simple HTML5 Android app and let it run on my smartphone.

`speechConfig.setProperty(SpeechSDK.PropertyId.SpeechServiceConnection_TranslationVoice, languageTargetOptions.value);`

The VoiceOut option is checked. That works fine so far. When I make a short break of speaking into the microphone the code plays back the audio of the translated input as it should. But the microphone still listens to the own speakers and trys to translate that "new" input again. So a loop of translations of translations [...] starts.
The simple solution would be to disable the microphone while a translated audio is played back. It is easy to get the events of starting play back with 

 `if (SpeechSDK.ResultReason[e.result.reason] == "SynthesizingAudio") {   }`

and of ending the play back with

 `if (SpeechSDK.ResultReason[e.result.reason] == "SynthesizingAudioCompleted") {   }`

to get the correct timings of disabling/enabling the microphone but I found no way how to do so. Stopping and re-starting the TranslationRecognizer is not solving the issue because stopping it means also to stop catching the events and I don't realize when the play back has finished. A workaround could be to grab the audio stream size in kilobyte and calculate the runtime of that audio clip. Then I could indeed stop the TranslationRecognizer and re-start it after a pause of the calculated time. But.... really, I do not like this solution. :) I have a feeling that this could bring other (timing) issues to my code. Just disabling and re-enabling the mic input would be much smarter. But how?

Best regards
Marc

**Comment by rhurey** [comment]
There's currently not a great way to do this that doesn't have some side effect that's less than optimal.

The perfect solution is for the device's audio system to have echo canceling where the microphone simply won't pickup the audio being played back.

The SynthesizingAudio events tell you when the service has returned the audio to you, it doesn't help you know when the playback device (if there is one, the end application could be writing the audio to a file, storing it...) finishes rendering the audio.

You're right that stopping and starting the recognizer will stop all events, but it has the advantage of also dropping any 1/2 sent phrases.

Another option would be to wrap the microphone in a custom stream that you could "pause" the data from the microphone by instead sending silence to the recognizer when speech was being played back and drop the microphone data yourself. This puts the cost of finding the microphone and implementing the microphone code on you, so that's not desirable. Then they'll be a timing issue around the potential for additional audio having been sent to the Speech Service before the synthesis for the current translation comes back.

In the end, for translation where the same microphone / speaker pair will be playing back the translated audio stream recognizeOnce is a better model than continuous recognition. This way you get one phrase of translation and it's audio at a time, and can restart recognition once playback is finished.

**Comment by Marcophono2** [comment]
Thank you very much, Ryan! You put all important points together. Yes, I think I will wrap the microphone access into a "native code" (as far as one can use this expression for a html5 app ;) to have direct access to the audio stream. Thanks again, Ryan!

Best regards
Marc

**Comment by wolfma61** [comment]
closing - if you have a related problem, please open a new issue
